### DISCUSSION TITLE
How to write an inverted related query?

### DISCUSSION THREAD

**Discussion by montyclt** [ravendb/ravendb#13922] [Q&A] [answered]
Hello, I have this documents:

**Customer:**
```json
{
  "Id": "1-A",
  "Name": "John Doe"
  [...]
}
```

**Contract:**
```json
{
  "Id": "2-A",
  "Name": "Main contract",
  "CustomerId": "1-A"
  [...]
}
```

Invoice:
```json
{
  "Id": "3-A",
  "ContractId": "2-A",
  [...]
}
```

In this point, I need to write a query to get a contract profile with information about the customer and invoices, by the contract ID:

Query contract profile with ID: 2-A
```json
{
  "Name": "Main contract",
  "CustomerName": "John Doe",
  "Invoices": [{...}, {...}]
}
```

I know I can use `let customer = RavenQuery.Load<Customer>(contract.CustomerId)` because I have the customer ID in my document, but how to do the inverted relationship? when is the other document (o multiple documents) that has a field with ID of queried document. I tried with LINQ `join into`, but is not implemented.

For me, is irrelevant if I can't do using LINQ and I need to use DocumentQuery o raw RQL, or I need to create an index, I only need an example of querying this type of relationship.

**Comment by ashokgudur** [comment]
The following documentation may help how RavenDB works.
https://bit.ly/DocumentRelations
https://bit.ly/RavenDbLoadingDocs

HTH

**Comment by ayende** [comment, marked as answer]
The typical way you'll do that is via a multi-map/reduce index. You'll have a map for `Contract` and `Orders` that merge them all together.

Something like:

```
// maps
from contract in docs.Contracts
let customer = LoadDocument<Customer>(contract.CustomerId)
select new { 
   CustomerName = customer.Name,
   ContractName = contract.Name,
   Orders = new object[0],
   ContractId = contract.Id
}

from order in docs.Orders
select new
{
   order.ContractId,
   Orders = new[]{new {order.OrderedAt, order.Value }} ,
   ConstractName = (string)null,
   CustomerName = (string)null,
}

// reduce
from result in results
group result by result.ContractId
into g
select new 
{
   ContractId = g.Key,
   Orders = g.SelectMany(x=>x.Orders).OrderByDescending( o => o.OrderedDate),
   CustomerName = g.Select(x=>x.CustomerName).FirstOrDefault(x=>x != null),
   ContractName = g.Select(x=>x.ContractName ).FirstOrDefault(x=>x != null)
}

```

**Reply by montyclt** [reply]
Can I add a limit to orders in this example? I want to get the contract profile with only last 5 orders.

**Reply by ayende** [reply]
Sure, in the Reduce, do this:

```
// reduce
/...
   Orders = g.SelectMany(x=>x.Orders).OrderByDescending( o => o.OrderedDate).Take(5),
//...
```
